tests/lib/eZ/RichText/Converter/Render/TemplateTest.php

```diff
@@ -76,58 +76,84 @@ public function providerForTestConvert()
-    {
+    public function testConvert(
+        DOMDocument $inputDocument,
+        DOMDocument $expectedOutputDocument,
+        array $expectedRenderParams
+    ): void {
         $this->rendererMock->expects($this->never())->method('renderContentEmbed');
         $this->rendererMock->expects($this->never())->method('renderLocationEmbed');
 
-        if (!empty($expectedRenderParams)) {
-            $convertIndex = 0;
-            foreach ($expectedRenderParams as $index => $params) {
-                if (!empty($params['params']['content'])) {
-                    // mock simple converter
-                    $contentDoc = new DOMDocument();
-
-                    $xml = '<section xmlns="http://docbook.org/ns/docbook">';
-                    $xml .= $params['params']['content'];
-                    $xml .= '</section>';
-
-                    $params['params']['content'] = $xml;
-
-                    $fragment = $contentDoc->createDocumentFragment();
-                    $fragment->appendXML($xml);
-
-                    $contentDoc->appendChild($fragment);
-
-                    $this->converterMock
-                        ->expects($this->at($convertIndex++))
-                        ->method('convert')
-                        ->with($contentDoc)
-                        ->willReturn($contentDoc);
-                } else {
-                    $params['params']['content'] = null;
-                }
-
-                $this->rendererMock
-                    ->expects($this->at($index))
-                    ->method('renderTemplate')
-                    ->with(
-                        $params['name'],
-                        isset($params['type']) ? $params['type'] : 'tag',
-                        $params['params'],
-                        $params['is_inline']
-                    )
-                    ->willReturn($params['name']);
-            }
-        } else {
-            $this->rendererMock->expects($this->never())->method('renderTemplate');
-        }
+        [
+            $convertParameters,
+            $convertReturnValues,
+            $renderParameters,
+            $renderReturnValues
+        ] = $this->provideConvertRenderValues($expectedRenderParams);
+
+        $this->converterMock
+            ->expects($this->exactly(count($convertReturnValues)))
+            ->method('convert')
+            ->withConsecutive(...$convertParameters)
+            ->willReturnOnConsecutiveCalls(...$convertReturnValues);
+
+        $this->rendererMock
+            ->expects($this->exactly(count($renderReturnValues)))
+            ->method('renderTemplate')
+            ->withConsecutive(...$renderParameters)
+            ->willReturnOnConsecutiveCalls(...$renderReturnValues);
 
         $outputDocument = $this->getConverter()->convert($inputDocument);
 
         $this->assertEquals($expectedOutputDocument, $outputDocument);
     }
 
+    private function provideConvertRenderValues(array $expectedValues): array
```

Comment:
Here also documentation of parameter/return value types is missing

Reply:
It might be tricky for this scenario, ref: https://github.com/ezsystems/ezplatform-richtext/blob/2.3/tests/lib/eZ/RichText/Converter/Render/TemplateTest.php#L161. Decided to leave it as-is.